### PR TITLE
Extensions compression

### PR DESCRIPTION
**Pull request by blunderer**

### CODE REVIEW

File: certificates.c [hunk 1]
```diff
@@ -79,53 +86,139 @@ hibacert_from_ext(struct hibacert *cert, struct hibaext *ext,
...
 int
-hibacert_load_extensions(struct hibacert *cert, struct sshbuf *extensions) {
-	int pos = 0;
+hibacert_load_extensions(struct hibacert *cert, struct sshbuf *blob) {
 	int ret = 0;
 	size_t i = 0;
-	size_t len = 0;
 	u_int32_t magic;
-	const char *data = NULL;
+	struct sshbuf *extensions = hibacert_maybe_inflate(blob);
```
Comment (burgerdev) [inline certificates.c:156]: `hibacert_maybe_inflate` has a bunch of branches that return `NULL`. Is it ok to ignore these here?

File: certificates.c [hunk 2]
```diff
@@ -79,53 +86,139 @@ hibacert_from_ext(struct hibacert *cert, struct hibaext *ext,
...
+	cert->exts = xreallocarray(cert->exts, cert->nexts + 1, sizeof(struct hibaext*));
+	cert->nexts++;
+	cert->exts[cert->nexts - 1] = hibaext_new();
+	ret = hibaext_decode(cert->exts[cert->nexts - 1], blob);
+	sshbuf_free(blob);
+	return ret;
+}
+
+static struct sshbuf*
+hibacert_maybe_inflate(struct sshbuf *extensions) {
```
Comment (burgerdev) [inline certificates.c:103]: I did not see corresponding deflate support, is that not implemented?

File: certificates.c (hunk 1 above)
Comment (blunderer) [reply certificates.c:156]: Done

File: certificates.c (hunk 2 above)
Comment (blunderer) [reply certificates.c:103]: Done. I converted the hibaext_encode_* to support multiple extensions merging implementing the compression, merging and base64 / raw encoding.

This code is exercised by the generate_test_certificates binary to confirm HIBA extensions serialized in any format can be properly deserialized.

File: extensions.h [hunk 3]
```diff
@@ -59,9 +74,17 @@ void hibaext_free(struct hibaext *ext);
...
-int hibaext_encode(const struct hibaext *ext, struct sshbuf *blob);
+/* Encode one or more HIBA extensions into a serialized blob to be included in a
+ * certificate (either raw or base64 encoded).
+ * blob must be already allocated, and will be reset.
+ * It is invalid to request encoding of more than 1 identity extension, since a
+ * host cannot expose multiple identities.
+ *
+ * The default hibaext_encode points to the base64 version, uncompressed single
+ * extension for backward compatibility. */
+#define hibaext_encode(ext, blob) hibaext_encode_b64((const struct hibaext**)&ext, 1, 0, blob)
```
Comment (burgerdev) [inline extensions.h:85]: The reference-taking in the macro feels a bit dodgy. Why don't we expose `hibaext_encode_one_raw` as `hibaext_encode`?

Nit: the `ext` might need parentheses.

Comment (blunderer) [reply extensions.h]: Instead I made the  hibaext_encode a small wrapping function (avoids increasing the breadth of entry points).

**Comment by blunderer** [comment]
The whole branch was rebased on top of another change that enforce -ansi at compile time. This makes compilation on darwin toolchain successful. The 4 commits from that pull request were mostly unchanged (except for the format of comments).

**Pull request merged**